Allow `zarr_path` to be of type `Path` in `echodata_combine`

b-reyes opened the pull request:
This PR addresses issue #841 by allowing the input `zarr_path` to be of type `Path` in `echodata_combine`. In addition to this change, I also made it  such that `zarr_path` is always a `Path` object in `zarr_combine` and thus I made it so that `check_zarr_path` always returns a `Path` object. 

From my preliminary investigation and local testing, there should be nothing preventing us from allowing `zarr_path` to be of type `pathlib.Path`. @lsetiawan I am putting you as a reviewer to confirm this as you are well versed in this type of topic.

lsetiawan commented on echopype/echodata/zarr_combine.py:
```diff
@@ -916,14 +917,14 @@ def _modify_prov_filenames(
         # obtain the filenames zarr array
         zarr_filenames = zarr.open_array(
-            zarr_path + "/Provenance/filenames", mode="r+", storage_options=storage_options
+            zarr_path / "Provenance/filenames", mode="r+", storage_options=storage_options
```
Seems like you're forcing `zarr_path` to be pathlib.Path only? How about if I want to pass in `s3://my-bucket/data.zarr`? casting that to `Path` will make that not work, turning the string to `s3:/my-bucket/data.zarr`.

lsetiawan commented on echopype/echodata/zarr_combine.py:
```diff
@@ -936,7 +937,7 @@ def combine(
         Parameters
         ----------
-        zarr_path: str
+        zarr_path: Path
```
I think this would push us backward to no cloud capability if it only accept `Path`.

lsetiawan commented on echopype/tests/echodata/test_zarr_combine.py:
```diff
@@ -237,7 +238,7 @@ def test_append_ds_list_to_zarr(append_ds_list_params):
     # create temporary directory for zarr store
     temp_zarr_dir = tempfile.TemporaryDirectory()
-    zarr_path = temp_zarr_dir.name + "/" + zarr_name
+    zarr_path = Path(temp_zarr_dir.name + "/" + zarr_name)
```
Please add a test that does same thing on minio.

b-reyes replied:
@lsetiawan I believe that my most recent changes make it unnecessary to add a test on minio. Please confirm.

lsetiawan replied:
We'll just revisit this addition when working in making more tests. I don't remember if there's a zarr_combine test that hit minio.

lsetiawan commented on echopype/echodata/combine.py:
```diff
@@ -65,20 +69,25 @@ def check_zarr_path(
 
+    # convert created validated_path to a Path object
+    if zarr_path is None:
+        validated_path = Path(validated_path)
```
This will break a path that's `s3://my-bucket/data.zarr`.

lsetiawan replied:
`validate_output_path` already deals with the complexity of `Path`. The downstream should just take the regular string of `validated_path`.

b-reyes replied:
I see how my implementation has issues. Thank you for pointing this out. In your opinion, should I make it so that `check_zarr_path` always outputs a string? From your comments, this seems like the best solution.

lsetiawan replied:
Yepp, `check_zarr_path` should always output a string. Thanks for working on this!

b-reyes replied:
Sounds good! I will go ahead and implement this.

lsetiawan commented on echopype/echodata/combine.py:
```diff
@@ -47,13 +47,17 @@ def check_zarr_path(
-        # check that zarr_path is a string
-        if not isinstance(zarr_path, str):
-            raise TypeError(f"zarr_path must be of type {str}")
+        # ensure that zarr_path is a Path object, throw an error otherwise
+        if isinstance(zarr_path, str) or isinstance(zarr_path, Path):
+            zarr_path = Path(zarr_path)
+        else:
+            raise TypeError(
+                "The provided zarr_path input must be of type string " "or pathlib.Path!"
+            )
```
```suggestion
        # ensure that zarr_path is a Path object, throw an error otherwise
        if not isinstance(zarr_path, (str, Path)):
            raise TypeError(
                "The provided zarr_path input must be of type string " "or pathlib.Path!"
            )
```

lsetiawan commented on echopype/echodata/combine.py:
```diff
@@ -47,13 +47,17 @@ def check_zarr_path(
         # check that the appropriate suffix was provided
-        if not zarr_path.strip("/").endswith(".zarr"):
-            raise ValueError("The provided zarr_path input must have '.zarr' suffix!")
+        if not (zarr_path.suffix == ".zarr"):
```
```suggestion
        if not str(zarr_path).strip("/").endswith(".zarr"):
```

b-reyes commented:
@lsetiawan I have changed the code so that `check_zarr_path` always returns a string. Additionally, I changed all associated path constructions so that they use `os.path.join` now, so that we can be OS independent. This PR is ready for another review.

lsetiawan commented on echopype/echodata/zarr_combine.py:
```diff
@@ -916,7 +917,9 @@ def _modify_prov_filenames(
         # obtain the filenames zarr array
         zarr_filenames = zarr.open_array(
-            zarr_path + "/Provenance/filenames", mode="r+", storage_options=storage_options
+            os.path.join(zarr_path, "Provenance", "filenames"),
```
This is a good idea, however I'm concerned that if the `zarr_path` is a url base to a bucket, there will be a problem on windows machine. `s3://my-bucket/data.zarr` can become `s3://my-bucket/data.zarr\Provenance\filenames`. I feel like there should be a function to handle this within your class.

b-reyes replied:
You are correct, I just tried it out on a Window's machine and the output is not correct. It will output the following: `s3://my-bucket/data.zarr\\Provenance\\filenames`. Do you have a function that already handles this? If not, I guess I can just create a function that determines if `s3://` is at the beginning of `zarr_path` and if it is concatenate appropriately. Do you know a better way to do this?

b-reyes replied:
The problem with my solution is that it limits us to `s3`. I am not sure what the other paths look like for other cloud storage options.

lsetiawan replied:
Anything with storage stuff will always have `://` so you can check for that in the path string and adjust as appropriate for the path concatenation. Currently we don't have a function that handles this. `sanitize_file_path` is the closest one: https://github.com/OSOceanAcoustics/echopype/blob/b58d1eb58dcc6dc33e26d196d809ae55c2ad3155/echopype/utils/io.py#L120. You can do similar things.

b-reyes replied:
I suppose we could just check for the main cloud storage options: AWS is `"s3://"`, Azure is `"az://"`, and Google Cloud storage is `"gs://"` (at least from what I can find).

b-reyes replied:
Ah, for some reason your comment did not show up! I can go ahead and create my own function using the logic `if "://" in file_path:`. Thank you for the suggestion.

b-reyes commented:
@lsetiawan thank you for your helpful comments. As you suggested, I have constructed a function named `join_paths` in `echopype/utils/io.py` that joins paths taking into account paths that are for cloud storage. I then changed all `os.path.join` occurrences to this function. This PR is ready for another review.

lsetiawan reviewed (changes requested):
Looks great! Thanks for incorporating those changes and for the new function! Very slick 😎 One more thing though, could you add a small test for the join_paths function? Almost there!

lsetiawan commented on echopype/utils/io.py:
```diff
@@ -280,3 +280,30 @@ def check_file_permissions(FILE_DIR):
+
+    Parameters
+    ----------
+    *args: str
```
```suggestion
    *args: tuple of str
```

b-reyes commented:
> One more thing though, could you add a small test for the join_paths function? Almost there!

@lsetiawan Do you know of a reliable way to create a test for this function accounting for the other possible OS? I looked into mocking, but this looks like it would not work.

b-reyes commented on echopype/utils/io.py:
```diff
@@ -280,3 +280,30 @@ def check_file_permissions(FILE_DIR):
         raise PermissionError("Writing to specified path is not permitted.")
+
+
+def join_paths(*args: str) -> str:
```
Good catch. Could you also add the `Tuple` import?

b-reyes commented:
@lsetiawan as we discussed, I have added another test that tests the true output of `join_path`. This test is OS dependent and the Windows associated parameters will be skipped if the test is not being run on a Windows machine. Please review that added code and run all io tests on a CI workflow using Windows.

leewujung commented on echopype/echodata/combine.py:
```diff
@@ -65,13 +65,18 @@ def check_zarr_path(
         save_path=zarr_path,
     )
 
+    # convert created validated_path to a string, if necessary
```
```suggestion
    # convert created validated_path to a string if it is in other formats, since fsspec only accepts strings
```

leewujung commented on echopype/test_data/README.md:
```diff
@@ -11,8 +11,6 @@ Most of these files are stored on Git LFS but the ones that aren't (due to file
 - 2019118 group2survey-D20191214-T081342.raw: Contains 6 channels but only 2 of those channels collect ping data
 - D20200528-T125932.raw: Data collected from WBT mini (instead of WBT), from @emlynjdavies
 - Green2.Survey2.FM.short.slow.-D20191004-T211557.raw: Contains 2-in-1 transducer, from @FletcherFT (reduced from 104.9 MB to 765 KB in test data updates)
-- raw4-D20220514-T172704.raw: Contains RAW4 datagram, 1 channel only, from @cornejotux
-- D20210330-T123857.raw: do not contain filter coefficients
```
Make sure you commit only the necessary files -- test_data/README.md should not be changed in this PR.

leewujung commented on echopype/tests/echodata/test_zarr_combine.py:
```diff
@@ -237,7 +239,7 @@ def test_append_ds_list_to_zarr(append_ds_list_params):
     # create temporary directory for zarr store
     temp_zarr_dir = tempfile.TemporaryDirectory()
-    zarr_path = temp_zarr_dir.name + "/" + zarr_name
+    zarr_path = join_paths(temp_zarr_dir.name, zarr_name)
```
For all uses of `join_paths` in this test function `test_append_ds_list_to_zarr`:
- the original form using pure string operation is not a good practice -- should use os.path
- since this test is to test the functionalities of combining datasets into zarr, using `join_paths` here introduces an unintended "test" on the path handling functionality of that function, making the tests hard to follow if anything breaks

I suggest removing the use of `join_paths` here and use os.path for creating the paths you need in this test.

leewujung replied:
One thing I am not sure is if os.path gives you `\` instead of `/` on Windows, is this why you used pure string operations? It seems that using PureWindowsPath you can "simulate" how the string should be interpreted in windows. How about using that?
Below from the 3rd block under this [Operators](https://docs.python.org/3/library/pathlib.html#operators) section:
```python
>>> p = PureWindowsPath('c:/Program Files')
>>> str(p)
'c:\\Program Files'
```

b-reyes replied:
As you have suggested, I have replaced the usage of `join_paths` with `os.path.join` in all tests where it is not necessary.

leewujung commented on echopype/utils/io.py:
```diff
@@ -280,3 +280,30 @@ def check_file_permissions(FILE_DIR):
         raise PermissionError("Writing to specified path is not permitted.")
+
+
+def join_paths(*args: Tuple[str, ...]) -> str:
```
Currently this function is only used in `zarr_combine._modify_prov_filename()` and the test functions, and not anywhere else. The functionality is among those provided by `pathlib`. I suggest removing this function and adding a block in the existing function `validate_output_path` to handle the Windows path changes -- this seems a good way to ensure that the path differences are handled correctly everywhere.

b-reyes replied:
Based on prior discussions we will not remove `join_paths` and we will proceed with the consensus established in the [PR comment](https://github.com/OSOceanAcoustics/echopype/pull/879#issuecomment-1320580649).

leewujung reviewed (commented):
@b-reyes : I am done with my review. My overall comment is that I think the handling of path can be in a block under `validate_output_path` so that that function can be reused everywhere. I suggest removing `join_paths` because its functionality is provided by pathlib and the function is only used once in the main codebase.

lsetiawan reviewed (commented):
I am no longer assigned as reviewer.

b-reyes commented:
Based on face-to-face discussions with @leewujung we have arrived at a consensus on how the items discussed in this PR should be addressed. The following is that consensus:
- The `join_paths` function should be kept and used throughout the code except for in tests that do not directly test the function. Additionally, @leewujung is considering changing the name of the function to something that is more specific. 
- To address the Windows issues we will take the following approach: At the beginning of `validate_output_path` if we are on Windows a string path input (that is not a cloud storage path) will be converted to `pathlib.WindowsPath`. This will allow `validate_output_path` to work properly and it will output a nicely formatted string that works with `os.path.join`. We can then proceed using `join_paths` as we currently are. 
- This discussion has also brought to light that other parts of the codebase may not be correctly constructing paths (when joining paths). For this reason, in this PR we should also review the codebase and correct how paths are joined.

lsetiawan commented:
There are some changes in this PR with `test_utils_log.py` due to some issues with `tempfile` trying to cleanup in Windows. This seems to be a common problem (See https://www.scivision.dev/python-tempfile-permission-error-windows/). A solution will be needed in a separate PR regarding any use of tempfile for creating temporary directory.

b-reyes commented:
@leewujung I have completed all of the items discussed in the above [comment](https://github.com/OSOceanAcoustics/echopype/pull/879#issuecomment-1320580649). One note:
- We had originally talked about replacing the below code with `validate_output_path`.
https://github.com/OSOceanAcoustics/echopype/blob/f92f9a0d8bc9af918de2fe2ece3880bd0d188e8d/echopype/convert/parsed_to_zarr.py#L41
However, after further review I found that this is not necessary. As it currently stands, we are working with a path that is created with `pathlib.Path.cwd` and all appending is handled appropriately with `pathlib`. The only reason we would need to modify this code section is if we allowed the user to choose where the temporary output produced by `offload_to_zarr=True` was stored. I think there is merit in allowing the user to choose the location of this  directory, but this enhancement should be addressed in another issue as it requires further investigation. If you do see merit in allowing the user to provide a path, please let me know and I will create an issue for this.

leewujung commented:
@b-reyes : I think this PR is ready to be merged. The only question I have is that `join_paths` is only used in the Provenance group attribute generation (and I remember you said that in other places, `xr.to_zarr/to_netcdf` takes care of this path appending issue already). I think it is fine to have it as a separate function for later reuse, but I gave the naming a bit of thoughts below:
How about changing the name to `append_group_path`? I do agree that this is not an appropriate name if we expand the use of this function in the future into actually adding actual path components, instead of path to a group within a Zarr store (the current use), but with the current use of this function, this name seems would avoid ambiguity, particularly wrt to those similar functions os.path and pathlib provides. Also, who knows, maybe pathlib would allow handling cloud paths soon and we don't have to worry about it anymore!

On the above comment re. `check_file_permissions` vs `validate_output_path`: we have agreed that we will keep the code as is for now, and I will add an issue for asking user to provide the final intended zarr path if they choose to use `offload_to_zarr=True`, both for efficiency and better practices on the cloud.

b-reyes commented:
> How about changing the name to append_group_path? I do agree that this is not an appropriate name if we expand the use of this function in the future into actually adding actual path components, instead of path to a group within a Zarr store (the current use), but with the current use of this function, this name seems would avoid ambiguity, particularly wrt to those similar functions os.path and pathlib provides. Also, who knows, maybe pathlib would allow handling cloud paths soon and we don't have to worry about it anymore!

@leewujung I am fine with changing `join_paths` to `append_group_path`. When we allow the user to specify the final intended zarr path if they choose to use `offload_to_zarr=True`, then we will likely not need to use this function because the full path will already be constructed and xarray or dask will handle the creation of the groups.

b-reyes commented:
@leewujung based on our offline discussion we will rename `join_paths` to `env_indep_joinpath` as this is the most descriptive and allows for future use in other places not tied to a zarr group.
